pkg/cloud/gcp/gcs_kms.go

```diff
@@ -0,0 +1,187 @@
+	params := kmsURIParams{
+		credentials: kmsURI.Query().Get(CredentialsParam),
+		auth:        kmsURI.Query().Get(cloud.AuthParam),
+	}
+
+	return params
+}
+
+// MakeGCSKMS is the factory method which returns a configured, ready-to-use
+// GCS KMS object.
+func MakeGCSKMS(uri string, env cloud.KMSEnv) (cloud.KMS, error) {
+	if env.KMSConfig().DisableOutbound {
+		return nil, errors.New("external IO must be enabled to use GCS KMS")
+	}
+	kmsURI, err := url.ParseRequestURI(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract the URI parameters required to setup the GCS KMS session.
+	kmsURIParams := resolveKMSURIParams(*kmsURI)
+
+	// Client options to authenticate and start a GCS KMS session.
+	// Currently only accepting json of service account
+	var credentialsOpt []option.ClientOption
+
+	switch kmsURIParams.auth {
+	case "", cloud.AuthParamSpecified:
+		if kmsURIParams.credentials == "" {
+			return nil, errors.Errorf(
+				"%s is set to '%s', but %s is not set",
+				cloud.AuthParam,
+				cloud.AuthParamSpecified,
+				CredentialsParam,
+			)
+		}
+
+		// Credentials are passed in base64 encoded, so decode the credentials first
+		credentialsJSON, err := base64.StdEncoding.DecodeString(kmsURIParams.credentials)
+		if err != nil {
+			return nil, err
+		}
+
+		credentialsOpt = append(credentialsOpt, option.WithCredentialsJSON(credentialsJSON))
+	case cloud.AuthParamImplicit:
+		if env.KMSConfig().DisableImplicitCredentials {
+			return nil, errors.New(
+				"implicit credentials disallowed for gcs due to --external-io-implicit-credentials flag")
+		}
+		// If implicit credentials used, no client options needed, just context
+		// So credentialsOpt is empty
+	default:
+		return nil, errors.Errorf("unsupported value %s for %s", kmsURIParams.auth, cloud.AuthParam)
+	}
+
+	ctx := context.Background()
+
+	kmc, err := kms.NewKeyManagementClient(ctx, credentialsOpt...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &gcsKMS{
+		kms:                 kmc,
+		customerMasterKeyID: strings.TrimPrefix(kmsURI.Path, "/"),
+	}, nil
+}
+
+// MasterKeyID implements the KMS interface.
+func (k *gcsKMS) MasterKeyID() (string, error) {
+	return k.customerMasterKeyID, nil
+}
+
+// Encrypt implements the KMS interface.
+func (k *gcsKMS) Encrypt(ctx context.Context, data []byte) ([]byte, error) {
+	// Optional but recommended by GCS
+	crc32c := func(data []byte) uint32 {
+		t := crc32.MakeTable(crc32.Castagnoli)
+		return crc32.Checksum(data, t)
+	}
+	plaintextCRC32C := crc32c(data)
+
+	encryptInput := &kmspb.EncryptRequest{
+		Name:            k.customerMasterKeyID,
+		Plaintext:       data,
+		PlaintextCrc32C: wrapperspb.Int64(int64(plaintextCRC32C)),
+	}
+
+	encryptOutput, err := k.kms.Encrypt(ctx, encryptInput)
+	if err != nil {
+		return nil, err
+	}
+
+	// Optional, but recommended by GCS
+	// For more details on ensuring E2E in-transit integrity to and from Cloud KMS visit:
+	// https://cloud.google.com/kms/docs/data-integrity-guidelines
```

Comment:
neat!

A brief read and these two points stood out:
```
If a decryption operation results in mismatched checksums, design your application to retry the operation a limited number of times, in case of transient problems.
If an encryption operation results in mismatched checksums for the encrypted data (the ciphertext), discard the encrypted result and try again.
```

I don't think we need to do this right away, but might be worth adding some exponential backoff retry behaviour if we see this error popping up frequently in the future.